### ISSUE TITLE
Implement NNEDI3 with GPU backend

### ISSUE THREAD

**Issue by maletor** [mpv-player/mpv#2230]
NNEDI3 works with a Vapoursynth port of the original script, but you must use `vspipe --y4m script.vpy >> output.raw` and then take a long walk. It would be amazing to implement NNEDI3 into MPV in such a way that it used the GPU through opencl or otherwise.

**Comment by haasn** [comment]
It's highly doubtful that we'll ever see OpenCL support in vo_opengl, unless it can be kept completely separate from the current code.

It might be possible to reimplement NNEDI3 in GLSL, but I have no immediate plans in doing so. (Plus, I'd first have to learn OpenCL and decipher cryptic “optimized” code to figure out how it works)

But maybe somebody else who cares more about the deinterlacer could take this up.

**Comment by bjin** [comment]
@maletor Just a side note. You don't have to use `vspipe` for this purpose, you can use `vapoursynth` video filter in `mpv` directly.

**Comment by maletor** [comment]
@bjin, you are correct. However, I will have to do some benchmarking to see how practical that is in real time. Does mpv offer any feedback on the milliseconds it took to render a frame?

**Comment by haasn** [comment]
Not in a user-friendly way, but you can disable framedrop and see how fast A/V gets out of sync to get some ballpark figures.

Most likely vapoursynth + CPU NNEDI3 will be like an order of magnitude too slow for realtime purposes, so measuring it is a bit moot.

**Comment by maletor** [comment]
@haasn, the image doubler is pretty powerful. There isn't anything much better than it even today and I think NNEDI3 first appeared in 2009, although these things are subjective by nature. Regardless, I think it should be highly desirable to get support in `mpv`.

Would be useful to see if an NNEDI3 port in GLSL already exists given how old it is.

**Comment by haasn** [comment]
Previous iterations of NNEDI and EEDI are even older than that, going back to 2005 at least. That said, I still doubt there is a port already - and I don't think NNEDI3 is as wide-spread as you make it out to be.

Apart from the original avisynth filter (which is what the VS filter is a port of), and the eventual OpenCL rewrite (which is what madVR uses), I don't know of any other versions.

**Comment by avih** [comment]
While I'm unfamiliar with NNEDI3, it sounds like one of its features (the major one?) is GPU based deinterlacer.

In general, I think mpv could benefit from GPU deinterlacer which doesn't depend on the GPU drivers or software (yadif).

That being said, while yadif is not perfect, it's also not terrible, and also supports frame "doubling" reasonably well. For non-videophiles users, it should be enough.

**Comment by maletor** [comment]
Remind me why we don't want OpenCL in `vo_opengl`?

On the other hand, it doesn't sound like it would be too hard to implement OpenCL isolated from opengl, perhaps as its own video output.

**Comment by haasn** [comment]
> In general, I think mpv could benefit from GPU deinterlacer which doesn't depend on the GPU drivers or software (yadif).

Agreed. Unfortunately, I'm not up to speed with the state of the art when it comes to deinterlacing, or more broadly, IVTC. It would definitely be useful to have an advanced, high-quality pulldown-aware, inter-frame deinterlacing IVTC algorithm directly in vo_opengl (could even easily re-use the existing interpolation code for this), but given that this is only really necessary for older TV broadcasts and stuff, you can also just easily do that offline with the plethora of tools available for the job (like vapoursynth).

> Remind me why we don't want OpenCL in ```vo_opengl```?

I'll let @wm4 answer that question.

> On the other hand, it doesn't sound like it would be too hard to implement OpenCL isolated from opengl, perhaps as its own video output.

It wouldn't need to be its own video output, just a separate .h/.c file exposing a simple interface (eg. an init, an uninit and a filter function). Most importantly, it should be possible to completely disable this at compile time. (And the one line in vo_opengl that would depend on it could be behind an #ifdef)

**Comment by wm4** [comment]
>Remind me why we don't want OpenCL in vo_opengl?

OpenCL is not known for stability, or even just stable performance. We'd have to maintain it, and it's going to cause LOTS of maintenance. OpenGL alone is already enough work.

Plus I'm afraid that it'd make the rest of the OpenGL code harder to maintain, because of course OpenGL and OpenCL will have to interact.

**Comment by maletor** [comment]
This isn't a huge priority. It's arguable the NNEDI3 filter adds too much stuff that isn't there and @haasn is probably right, vapoursynth support should be more than enough for now.

**Comment by bjin** [comment]
I was searching for other deband shaders the other day and found [this repo](https://github.com/zachsaw/MPDN_Extensions/tree/master/Extensions/RenderScripts), it includes some HLSL shaders that MadVR [provides](http://www.videohelp.com/software/madVR/version-history#changelog) and is licensed under LGPL 3.0, like nnedi3/SuperRes/Super-xbr. Potentially they can be ported to GLSL (I'm not sure since I know absolutely nothing about HLSL), but my question now is how likely could these be integrated into mpv, considering the maintenance cost.

If they are only supposed to be supported via user provided shaders, the currently custom shader interface might not be enough AFAIU. Some shaders might require multiple passes, and after each pass the texture might be transformed as well. The width/height might be doubled, the texture might be rotated by 90 degrees, also a half pixel shifting could be introduced. For shader like nnedi3, it would be ideal that we can use mpv built-in scaler to do the final upscaling and correct the center shift within one pass.

@haasn

**Comment by haasn** [comment]
The only thing that the current shader API is inadequate for is shaders requiring multiple passes and shaders changing the texture size; and for both of those I'm still taking suggestions on how to design an API to handle this.

**Comment by bjin** [comment]
A straightforward way would be leaving the texture transform parameters (with appropriate tags)  in shader comments, but probably not a nice approach.

Another trickier issue is that nnedi3 shader requires data from an external binary to work. Eventually these data can be embedded into the shader, but having mpv to read these data and pass them as another texture would be nice.

**Comment by wm4** [comment]
Some possible choices...

1. Don't design an API and make it builtin
2. Make an overly specific API, which is designed to handle exactly this case
3. Add full scripting language support

**Comment by haasn** [comment]
> A straightforward way would be leaving the texture transform parameters (with appropriate tags) in shader comments, but probably not a nice approach.

This is not as simple as you make it sound, I've discussed reasons for this over at https://github.com/mpv-player/mpv/issues/1945#issuecomment-109941415

> Another trickier issue is that nnedi3 shader requires data from an external binary to work. Eventually these data can be embedded into the shader, but having mpv to read these data and pass them as another texture would be nice.

I assume you're talking about the NN weights? These could very well just be hard-coded constants in the shader source, though that might be slower than reading from a texture.

> 1. Don't design an API and make it builtin

This is definitely the easiest, for everything related. We could go ahead and start factoring out individual gl_video features to separate files, too, thus making the whole thing more modular in the process. There's nothing really stopping us from doing so right now.

> 3. Add full scripting language support

This would be the only other sensible option - perhaps it could be made so that you can set up the shaders via Lua or something? It would have to run on every single frame, though, due to the current design of vo_opengl - although it may be possible that we could build in even more caching mechanisms than we currently use. (eg. seeing if the vo config has changed at all, and if so, just repeating the previous passes without even generating GLSL)

**Comment by bjin** [comment]
IMHO, if we consider this a bit further, it probably makes more sense to have scripting language support for shaders. Not only because shaders other than NNEDI3 can be ported easily as well (with a glance at these HLSL files, it seems that super-xbr is easier to port), but also getting rid of the maintenance cost and the potential license issue.

**Comment by bjin** [comment]
I spent some time reading these HLSL shaders, and believe it's possible to port them to GLSL. Further more, if we are not going to take the generic approach (with scripting support), another option could be adding another kind of scaler different from current ones based on convolution. Basically these scalers will keep all known pixels and trying to interpolate the others, and as a result frame size will typically be doubled and requires a final convolution based scaler to reach the target resolution (and correct shifting). Currently this includes super-xbr/nnedi3/nedi shaders from MPDN extensions repo (SuperRes is a post processing filter and is pretty difficult to integrate). Any thoughts?

My proposal to the option name includes "pre-scaler", "double-scaler" or "iscaler".

**Comment by wm4** [comment]
Sounds ok in theory.

**Comment by haasn** [comment]
I'm not sure if NNEDI3 really keeps the original pixels the same. Does it really only write to every other pixel?

**Comment by bjin** [comment]
It's a deinterlacer, so that's how it's supposed to work. Given a interlaced footage, a non-temporal deinterlacer like nnedi3 will simply ignore one field and doing the interpolation for pixels on this field instead (based on the remaining field). If we pretend all pixels from current frame is from the same field, then we can keep all existing pixels and double the height of the frame.

**Comment by haasn** [comment]
Oh, yeah, that makes sense.

There's actually a special case here because it only deinterlaces/doubles in one direction at a time. So perhaps it might make more sense to have the double-shader only double the image in one direction (and simply apply it twice, with rotated vertices).

What about the others (SuperRes etc.), do they gain or lose anything by doing it in two unidirectional passes instead of one bidirectional pass?

**Comment by bjin** [comment]
Yes, for NNEDI3, it will be easier just to double height or width in different pass, but for quality it's  better to do it alternately (if we are going to do 4x upscaling).

Super-xBR and NEDI also requires two passes to double both the height and width of a frame, but taking a slightly different approach.

```
A B C
D E F
G H I
```

Here, `A C G I` are known, all other pixels are to be interpolated. So the idea for super-xbr and nedi is to interpolate `E` in the first pass, and then pretend image is rotated by 45 degrees and interpolate `B D F H` in the second pass.

SuperRes is some kind of post processing filter which attempts to eliminate artifacts in multiple iterations  with some heuristic methods, it seems to work well but too difficult to integrate into mpv.

**Comment by haasn** [comment]
> but for quality it's better to do it alternately (if we are going to do 4x upscaling).

Yeah, but these would still be separate invocations of each doubling pass. (ie. double W, double H, double W, double H, double W, double H, ...)

> Super-xBR and NEDI also requires two passes to double both the height and width of a frame, but taking a slightly different approach.

Okay, that would certainly not fit into the model I had in my head, which would be for orthogonal rotations only.

At this point I think it might just be easier to integrate NNEDI3, super-xbr, NEDI, SuperRes etc. directly into mpv, if a GPL'd source that is not dependent on OpenCL can be provided for them.

**Comment by bjin** [comment]
> Yeah, but these would still be separate invocations of each doubling pass. (ie. double W, double H, double W, double H, double W, double H, ...)
`double W` and `double H` can be the same, and we don't have to actually rotate texture by 90 degrees or 45 degrees, either. If we are going to generate shader code (instead of providing two glsl files), we can just change the parameter passed to `texture()`, or define a macro for that. In this way, most shader code will be identical for both doubling passes. The same also applies to NEDI and Super-xBR.

**Comment by haasn** [comment]
Yeah, obviously if we build it into mpv we can do whatever we want. I was just thinking of ways to expose a public API for shaders like these, and one way would be to have the shader always do vertical interpolation and simply change the vertices' texture coordinates such that for the “horizontal” pass it thinks it's doubling a HxW texture vertically, without requiring the texture to be rotated.

Any way, the important first step would be getting GLSL versions of these shaders, then we can think about integrating them.

**Comment by bjin** [comment]
Porting `super-xbr` seems to be quite easy, simple function/type name substitution will probably work. I also know how `NEDI` works (there is an [article](http://chiranjivi.tripod.com/EDITut.html) which is very to understand). But I didn't actually look into how `NNEDI3` works, the shader code appears to be clear though.

**Comment by bjin** [comment]
@haasn I ported `super-xbr`, see https://gist.github.com/bjin/1ad32f053f40d1a7795e

Change `#define Pass 0` to `#define Pass 1` for second pass, and then pass `--window-scale=2 --vo=opengl:scale=custom:scale-shader=super-xbr-p0.glsl:post-shaders=super-xbr-p1.glsl` to `mpv`.

It looks sharp, and with less aliasing, but also comes with noticable ringings.

I will probably work on `NEDI` next.

**Comment by maletor** [comment]
That was fast.

[This tool](http://sourceforge.net/projects/hlsl2glsl) possibly useful. It may prove to be a good starting point or may be prove to be more work than it's worth.

**Comment by bjin** [comment]
Forget to mention the shifting this shader introduced. The top left pixel stays at the top left corner after doubling, so the offset vector is `(-0.5, -0.5)`.

**Comment by haasn** [comment]
@wm4 I'd like move this to gl_video, what do we do about the license header in that script?

**Comment by wm4** [comment]
What license is the script?

**Comment by bjin** [comment]
The whole project is licensed under LGPL 3.0, but this shader alone seems to be licensed under MIT.

**Comment by haasn** [comment]
Concerning the license, I've decided to restructure gl_video.c to move individual features to separate files, this will just end up in its own file with the original header preserved (and the code relicensed under mpv's GPL/LGPL dual-license).

@bjin I noticed that the shader operates on RGB space but converts its input to YUV (or at least samples the brightness) before modifying it.

Since we have control over the color chain, we could also just simply run this shader before RGB conversion (if possible) and directly sample the Y plane, only falling back to the manual brightness calculation if the source video was RGB to begin with.

Do you know if having the algorithm itself run on images in the YUV space will affect its image quality in any meaningful way? Given the fact that it seems to be mostly doing linear operations like matrix multiplications and mixing, I'd guess that this shouldn't affect quality, but I don't understand the algorithm enough to be sure.

Also, what's the deal with FAST_METHOD? Why does that constant exist?

**Comment by bjin** [comment]
I would suggest we leave it as it is, at least for now.

1. It's actually badly named, change it(`RGBtoYUV()`) to `luma()` would be better.
2. The quality gain this change bring is minimal, the luma channel is only used to calculate weights (`d_edge` and `hv_edge`), it doesn't help much if we improved accuracy of luma channel.
3. I suppose this shader will somehow work on YUV, after we get `d_edge` and `hv_edge` with luma channel, the left work is basically `mix` and `clamp` on existing pixels. Although I could imagine some minor changes are necessary to do it properly.
4. It doesn't help improving the performane, it's not the bottleneck. Eventually we need to do YUV to RGB conversion, which is slower than extracting luma channel from RGB. Furthermore, doing YUV to RGB conversion on a higher resolution actually could hurt the performance, it's not the bottleneck either, though.

It's interesting since eventually we need to consider using these shaders on colorspace other than RGB, if we plan to improve the quality of chroma upscaling. But for know, stick it to RGB colorspace would be enough and less confusing.

> Also, what's the deal with FAST_METHOD? Why does that constant exist?

It seems to be a flag to improve the performance a bit, but it doesn't use the calculated `hv_edge` so I presume it's with some quality loss. My suggestion is to remove these routines entirely. The shader is already pretty fast as I tested.

As for `XBR_EDGE_STR` and `XBR_WEIGHT`, we could make them `param1` and `param2` and leave them undocumented for now.

**Comment by haasn** [comment]
I would prefer running on the actual luma channel for code quality reasons more than anything. Hard-coding arbitrary luma constants is not something I would be doing either way. If I have to fall-back, I would base it off the video's associated primaries. (It would also, of course, be faster to avoid the luma calculation)

I speculate that the real “bottleneck” of the shader is going to be the texture sampling, because vo_opengl already makes heavy use of it. So if anything is going to impact overall performance, that would most likely be it.

**Comment by bjin** [comment]
> Hard-coding arbitrary luma constants is not something I would be doing either way.

I see, that make sense. Let's assume there is no difference applying this shader directly on YUV, YUV to RGB is linear so if we ignore these `clamp` and `step` functions, the result should be identical.  The only concern I have now is RGB linearization, which will make the results significantly different.

**Comment by bjin** [comment]
Regarding the `NEDI` port, I don't plan to do it anytime soon now. I read the shader, it's quite different from the original algorithm. A very small window size is used, probably for performance purpose. And to prevent the overfitting it caused (as well as the other artifacts), some weird techniques are used which I don't understand. Also, in the revised version doubling of height and width are seperated, which I also have doubts with.

Anyway it shouldn't be hard to port, so probably anyone else will do it.

**Comment by bjin** [comment]
I think I now get some idea how the `NNEDI3` shader works. Unfortunately this shader can only process one plane (mostly luma) in a single pass, and this makes the integration quite difficult I believe.

Basically it takes luma value of samples from a local `4x8` window, and encode 4 samples (scalar type) from the same column into a vector for optimization purpose. We can, of course, expand the most inner loop to process multiple plane (like `RGB`) in single pass, but it will lead to performance degration of more than three times I believe. Considering `NNEDI3` is known for being slow, it might not be a popular option.

One way to resolve this is convert RGB back to YUV and process luma and chroma plane sepeartely in `NEEDI3` shader, this approach looks ugly but should have a good tradeoff between the performance and rendering quality.

Otherwise we need to introduce another option to make it possible to do the RGB conversion at the target resolution. For a typical subsampled YUV frame

* `Y` plane will be doubled with `NNEDI3`, and scaled to target resolution with `scale`, probably with shifting corrected.
* `UV` plane will be scaled to target resolution directly with `cscale`.
* colorspace transform happens at the target resolution.

**Comment by haasn** [comment]
> Unfortunately this shader can only process one plane (mostly luma) in a single pass, and this makes the integration quite difficult I believe.

The only potential hit is code quality. But I wanted to refactor the pre-RGB bits of the code either way to make this kind of stuff easier.

> We can, of course, expand the most inner loop to process multiple plane (like RGB) in single pass, but it will lead to performance degration of more than three times I believe.

GLSL is known for being able to perform operations in parallel on multiple channels. I'm not sure why that wouldn't apply here. Instead of a 4D vector for the column you'd have a 4x3 matrix.

> One way to resolve this is convert RGB back to YUV and process luma and chroma plane separately in NEEDI3 shader, this approach looks ugly but should have a good tradeoff between the performance and rendering quality.

This would only be an issue for sources that RGB originally, and in this case I would be happy simply disabling NNEDI3 support if it would be a big problem. (But see above)

> Otherwise we need to introduce another option to make it possible to do the RGB conversion at the target resolution. For a typical subsampled YUV frame

I don't like this approach; and I don't see why it would be necessary. Just perform luma doubling before the usual conversion to RGB.

**Comment by bjin** [comment]
> I don't like this approach; and I don't see why it would be necessary. Just perform luma doubling before the usual conversion to RGB.

Yes, this is another option. The concern regarding this is we still need to correct the shifting `NNEDI3` introduced before we do RGB conversion. But I admit it's really a minor issue.

> GLSL is known for being able to perform operations in parallel on multiple channels. I'm not sure why that wouldn't apply here. Instead of a 4D vector for the column you'd have a 4x3 matrix.

Okay, it's fine then. Another question, if we call `sample().x`, will the shader compiler be "smart" enough to reduce the work to `1/3` for us in this case?

**Comment by haasn** [comment]
> The concern regarding this is we still need to correct the shifting NNEDI3 introduced before we do RGB conversion.

Is this just an offset, or does it require stretching/rescaling?

Offsets are trivial to account for. Stretching requires work, although if it's only in one direction it can be done very fast - especially if it's only by a small amount. (Also, bilinear would be free either way)

> Another question, if we call sample().x, will the shader compiler be "smart" enough to reduce the work to 1/3 for us in this case?

In general, yes. Although note that sampling in general does not get a lot faster if you leave away channels - it only cuts down on the number of ALU operations you need to perform afterwards.

**Comment by bjin** [comment]
> Is this just an offset, or does it require stretching/rescaling?

It's just offset. I assume we still need another pass to use some convolution-based scaler (preferably the EWA based ones) to correct this offset. The only other option I can think of is leave it as is and let OpenGL's builtin bilinear resampler to handle that.

**Comment by bjin** [comment]
A [proof-of-concept GLSL port](https://gist.github.com/bjin/97c15910fd1f563bac18) of `nnedi3`.

* Test by passing `--vo opengl-hq:pre-shaders=\"luma.glsl,nnedi3.glsl\" --video-unscaled` to `mpv`
* It acts as deinterlacer now, top field will be kept. It should be easy to migrate to doubler later, by changing first few line to modify pixel coordinates.
* It can process only the luma plane, so I added another shader to extract the luma plane.
* Currently `nns` was set to 32, I also include other `.inc` files for NN weights. But as I tested, `nns=64` will fail the shader compilation, so we need to figure out another way to do this.

**Comment by haasn** [comment]
> It's just offset. I assume we still need another pass to use some convolution-based scaler (preferably the EWA based ones) to correct this offset.

Ah, yes, that would be true for offsets that are not whole integers (ie. offsets like 0.5), since they need to interpolate in-between pixels.

Bilinear for that is not very nice since it makes everything instantly very blurry.

Note that if the worry is offset between the luma and chroma planes, we could also just move the chroma plane (which is being upscaled with a convolution filter either way) over by the same amount as the luma plane gets shifted after doubling.

**Comment by haasn** [comment]
There's one more thing I'm not sure I understand about the super-xbr shader. How is the texture size increased supposed to work?

Say we have a 1280x720 RGB source as our input. As I understand it, the output of the second pass would be a 2560x1440 texture (more technically: I want to sample the second pass's shader at that resolution), but what about the first pass? Would it also output at 2560x1440?

**Comment by bjin** [comment]
> Would it also output at 2560x1440?

Yes. But only half of the pixels from the output of first pass is meaningful.

**Comment by haasn** [comment]
> Yes. But only half of the pixels from the output of first pass is meaningful.

Then perhaps as a later optimization it would be a good idea to only draw half of each texture per pass. This could either be done by using a conditional inside the shader, or perhaps some kind of checkerboard “mask” that restricts the region in which OpenGL draws.

**Comment by bjin** [comment]
> Then perhaps as a later optimization it would be a good idea to only draw half of each texture per pass. 

I'm not sure if this gonna help or not. The "garbage" pixels produced by first pass will be shortcut in the first pass, and will never be used in second pass. However, we could just return dummy pixels to save the extra sampling time.
```
#if Pass==0
    vec2 dir = fract(pos * tex_size) - 0.5;
    if (dir.x * dir.y < 0) return vec4(0, 0, 0, 0);
    if (dir.x < 0 || dir.y < 0) return texture(tex, pos - dir / tex_size);
#elif Pass==1
    ...
```

**Comment by haasn** [comment]
Okay, after playing around with the shaders a bit more and getting to understand the way it's masking and writing the pixels, I have come up with this alternative approach. For reference/explanation, I have some illustrations of the process:

![SuperXBR illustration](https://0x0.st/sFD.png)

Left is the original image. The middle image is the structure of the first pass (‘untouched’ pixels in their original color, blue pixels are the garbage pixels, and red pixels are the useful output). The right image is the structure of the second pass. Red pixels are left over (untouched) from the first pass, and the green pixels are the new pixels added in this pass.

It seems to me that we could very easily optimize this by simply having the first pass draw its images to a separate texture (“red” texture) instead. In other words, the first pass would sample from the 100x100 original and create a 100x100 texture containing only red pixels.

The second pass would then sample from both the original and red textures and draw out all of the green pixels (plus the original pixels where relevant).

Would require modifications to the code though, I'll have to figure it out (especially in terms of performance). Another thing I noticed is that it does a lot of transformations of pixel coordinates inside the shader - but we could “help” it out here by using the GPU's rasterizer instead.

But in both cases I think I will implement the algorithm as-is first so I have both a performance metric and a reference image to make sure my speedups do not change the algorithm result.

I'm mainly documenting this in case it helps you think about similar ideas for NEDI, NNEDI3 etc.

> if (dir.x * dir.y < 0) return vec4(0, 0, 0, 0);

Incidentally, I learned that you can use the GLSL keyword ‘discard’ to simply exit out of the shader without drawing anything at all. (They would be left in an undefined state, but since the ‘blue’ pixels are not read from this is a non-issue)

**Comment by haasn** [comment]
So, after implementing the rest of the algorithm it turns out there actually is a slight problem with the “blue pixel” optimization: due to the way texture clamping works, pixels that end up near the edges will end up sampling from “blue” regions because the corner pixels conceptually extend to infinity here.

Note that if we do the red pixel rework that I mentioned above, this would be a non-issue (because the red pixels and original pixels would get clamped separately, there are no “garbage pixels” to worry about).

Anyway, I'm not confident if the result is correct, because the results I'm getting look a bit terrible to be quite frank:

This is what I see on my screen now, roughly. Note that the “true” upscaling is still being done with nearest neighbour - but you can see nonetheless how the xbr step doubles the effective image resolution.

![](https://0x0.st/sFd.png)

Ignoring for a moment the blue fringes around the edges (which I already explained), does the rest of the image look as expected?

**Comment by haasn** [comment]
I found some example images of super-xbr on wikipedia, though it uses “Super-xBR 4x” instead of the 2x version I was implementing.

I can technically just chain the 2x version multiple times (in fact the code trivially supports this), but I'm not sure if they are using a different algorithm that's optimized for 4x or something.

Either way, here are my comparisons... there *are* some differences to the wikipedia versions. I'm not sure where they arise from, since I don't know what algorithm that was using. (Ignoring, again, the left-over blue pixels)

![](https://0x0.st/sFn.png)

Do you know of a better source I could use as a “reference” for this particular algorithm, to make sure I did not simply make a mistake while implementing it?

**Comment by haasn** [comment]
Another comparison, this time with the [madVR 4x screenshots[(http://forum.doom9.org/showthread.php?p=1726441#post1726441), using the “no AR” version (I also disabled AR in my version of the algorithm).

![](https://0x0.st/sF5.png)

There are very clear and obvious differences, leading me to believe that either this GLSL shader is faulty, my reimplementation in mpv is incorrect.

You can see my code at https://github.com/haasn/mpv/commit/8a29b18b6ee97f248b1b80fe20444e35d9f1ca85

**Comment by haasn** [comment]
Update: I found the issue I had in the code, I copied the wrong values for the constants near the top of the shader.

Now my results are identical to the results I get with your GLSL shader. Unfortunately, the issue with the blue pixels around the edges persists. I've also realized that super-xbr is an absolutely terrible filter, and thus I don't see much point in trying to optimize it either.

I'll just leave out the blue-pixel optimization for values near the edges.

**Comment by bjin** [comment]
> I'll just leave out the blue-pixel optimization for values near the edges.

My original revision uses nearest neighbour for blue pixels, it's just for testing purpose for the first pass. Maybe change it to bilinear might be proper.

Forcing the number of passes in vo might be unfriendly to user, the parameters I have in my mind are something like:
* `pre-scaler`: can be `none` or `super-xbr`.
* `pre-scaler-min-upscaling-ratio`: The minimum upscaling ratio requires to apply the `pre-scaler`, defaults to `1.4`.
* `pre-scaler-max-times`: The maximum times that `pre-scaler` are allowed to perform, defaults to `2`.
* `use-pre-scaler-for-chroma`: as described.

**Comment by haasn** [comment]
Okay, I've pushed my latest set of fixes and improvements to the branch. This is now almost in a state where I would be okay with merging it, the last thing I want to find out is if it would be better to do the scaling before or after RGB conversion, and before or after sigmoidization, respectively. (I also want to get rid of the hard-coded luma coefficients in the process.)

I'll continue working on this tomorrow.

> My original revision uses nearest neighbour for blue pixels, it's just for testing purpose for the first pass. Maybe change it to bilinear might be proper.

My current version ‘discards’ blue pixels except where they're near the edge, where it applies the normal super xbr algorithm.. but probably bilinear would be better here. I'll change that.

> pre-scaler: can be none or super-xbr.

I'd like to keep this conceptually separate from eg. NNEDI3, especially since we can very easily support both at the same time. (They also probably run at different times, so it would be confusing to have both use the same option)

> pre-scaler-min-upscaling-ratio: The minimum upscaling ratio requires to apply the pre-scaler, defaults to 1.4.

I don't think the VO should really be toggling options on and off like this. If anything, I would prefer it if this is achieved, after the planned redesign of the options system, by using a script or profile that sets options based on the source resolution (or scale factor). So I will not implement any “selection” logic like this for now.

> use-pre-scaler-for-chroma: as described.

I would prefer simply adding a new option to cscale, like cscale=superxbr to re-use the code here. (The design makes it easier enough)

Note that super xBR for cscale, unlike NNEDI3, has a potential problem: it can't scale 4:2:2 content at all, since that requires scaling in only one direction.

**Comment by bjin** [comment]
Regarding the `NNEDI3` port, I realized that it might be better just add outer-most loop to handle different channels, instead of expanding the inner-most loop the add ugly code. In this way, it's easier to reason about the performance, also to adopt the patches from upstream easier (This filter is still changing rapidly there).

Ideally, I still would like to have an option to apply `nnedi3` on luma plane only (My graphics card is crappy), and correct the offset by applying the same to chroma plane upscaling looks acceptable for me. Eventually we can revert this offset on RGB plane later.

For `YUV` frames:
* `Y` plane will be doubled with `NNEDI3` (probably multiple times), and introduced an `offset`
* `UV` planes will be scaled with `cscale` to the size of doubled `Y` plane, with `offset` **applied**
* `YUV` to `RGB` conversion
* `scale` (or `dscale`) applies on `RGB` frame, optionally with `offset` **corrected**

For `RGB` frames (or other formats)
* Applies `NNEDI3` on three channels, double the image and introduced an `offset`
* Applies `scale` on doubled frame, and with `offset` corrected

Also, things get trickier if we also plan to applies pre-scalers on chroma plane.

**Comment by haasn** [comment]
> Eventually we can revert this offset on RGB plane later if we want.

Great idea, I didn't even think of that. Note that we would still lose like 1 pixel near the edge of the screen but it wouldn't be too dramatic.

I think an important first step would be getting a working version of NNEDI3, then we can bench it and think about how to best fit it into mpv. Our code is fairly flexible.

**Comment by bjin** [comment]
> An [incomplete GLSL port](https://gist.github.com/bjin/97c15910fd1f563bac18) of `nnedi3`.

Just in case you missed it, I posted my `NNEDI3` port earlier. It's mostly usable.

1. If we want to double the resolution, we need to modify the pixel coordinate related routines, for both direction.
2. To support `RGB` planes, we need to add another outer-most loop, or wrap it in another function.

But I guess it's still requiring considerable amount of work, mainly on refactoring `mpv` code. The lucky part is that we have `nnedi3` vapoursynth plugins for reference purpose.  There is changes porting it to `OpenCL` but I presume it's only minor and doesn't hit the quality badly.

**Comment by haasn** [comment]
> Just in case you missed it, I posted my NNEDI3 port earlier. It's mostly usable.

Oh wow, I will definitely have to play around with this tomorrow. The code is much shorter and simpler than I anticipated, even simpler than super_xbr by some margin.

There is one thing that I'm curious about, though. As it stands, for each output pixel, the shader samples an 8x4 region of memory and.. does something to it.

Does this 8x4 region differ for each pixel, or is there some potential for work re-use here? If it would be beneficial to cache some common information out into a separate texture, we could very easily do that.

**Comment by bjin** [comment]
> Does this 8x4 region differ for each pixel, or is there some potential for work re-use here?

It's different for each pixel, basically these 32 samples from a local window are taken as input of NN process.

Regarding the NN weights (the five huge `.inc` files), since essentially an array of `vec4`, we could just pass three extra `GL_NEAREST` textures as `uniform`. My concern is, how we handle these weights, could we just commit them into `mpv` source tree directly?

**Comment by wm4** [comment]
How huge are they?

**Comment by bjin** [comment]
> How huge are they?

5 text files with `float` constants, ranging from `24KB` to `360KB`. If it's only about file size, we could encode them into binary (probably could save more than half), and use some python script to deserialize them during the build.

**Comment by wm4** [comment]
Link?

**Comment by bjin** [comment]
> Link?

https://gist.github.com/bjin/97c15910fd1f563bac18 (The `.inc` files)

**Comment by haasn** [comment]
I didn't realize they were that large.

I'm actually worried about the performance of this more than anything. Last time I tested using hard-coded float constants (in GLSL), I was using an array of 256 floats hard-coded into my source and it got unusably slow. At 64 or fewer floats it was better, however, with a larger array generally resulting in greatly reduced performance. (Compile-time was also very slow)

Sampling from a texture eliminates these concerns entirely, but sampling from a texture is also the thing we are trying to *avoid* doing when it comes to vo_opengl performance...

As a possible third option, we could potentially update a float array as a uniform. I forgot what kind of performance characteristics this had, but I think it was also quite poor compared to a texture.

I think due to these concerns a texture would still be the safest bet, so we would probably include it as some .h during compile time that will be used to generate the textures as needed. (Plus, if the user is using NNEDI3 doubling, they most likely aren't using very heavy correction scalers)

**Comment by bjin** [comment]
Another idea to handle these NN weights. We could ask the user to specify path to `nnedi3_weights.bin
` (provided [by the vapoursynth plugin](https://github.com/dubhater/vapoursynth-nnedi3/blob/master/src/nnedi3_weights.bin)). The latter file seems to be binary serialization of single precision `float` types.

**Comment by haasn** [comment]
What purpose would the user customization have?

Unless there is a licensing issue I see little reason to require a third party dependency on a binary file containing meaningless numbers that the user has zero justification to change.

**Comment by wm4** [comment]
Because it's fucking huge. The mpv binary would consist of more float numbers than actual code. It's not compressible either (with gzip or xz). Maybe some sort of special filtering could reduce it, but as I understand, the numbers are essentially random?

So yes, that's a problem.

I don't really see how the previously posted inc files amount to 13MB, though.

**Comment by bjin** [comment]
> I don't really see how the previously posted inc files amount to 13MB, though.

The original filter supports window size other than `8x4`, like `32x4` or even `48x6`.

**Comment by wm4** [comment]
From the inc file I counted approximately 2^15 floats, 128 KB.

**Comment by wm4** [comment]
Anyway, 128KB of data gets my blessing, 13MB not. 13MB is just too ridiculous.

(What I'm worried about is how to get these 128KB into git, though.)

**Comment by avih** [comment]
Could these constants be generated at runtime somehow? I find it hard to believe someone hand tuned 13M of data. Or is it a result of running some code on some samples set and finding out the constants which will perform best?

**Comment by wm4** [comment]
From what I heard, it's a trained neural network.

**Comment by haasn** [comment]
It's a trained neural network. Some notes are in order:

1. As I understand it, you can vary the number of neurons and the window size (typically 8x4) independently. While it would be cool for “bragging rights” if we could vary the window size as well, I don't think it justifies the size increase. That already significantly cuts down the majority of the size.

2. As I understand it, the number of neurons dramatically affects performance - with 256 neurons being so implausibly expensive that modern GPUs can't hope to run it in realtime. Given this realization, I would safely say we can limit it to what's reasonable in realtime.

If we limit it to the simplest form and realistic numbers of neurons, that should be much, much less data.

As for how to get it into git, I would just drop an X-line constant float array into its own file and import that (+ upload it as a texture).

**Comment by bjin** [comment]
I modified both `superxbr` and `nnedi3` shaders, and actually get them working in a testing environment. But as I'm worried, it's far from being ready to be merged (or reviewed).

The changes I made to shaders:
* Add proper two pass coordinate handling to `nnedi3`. It can actually be used as "upscaler" now.
* Unified the interfaces of `superxbr` and `nnedi3`. They are both meant to process one plane only now. This avoided the confusion when we trying to find the luma value from RGBA space for `superxbr`, and also make it easier to port both `superxbr` and `nnedi3`.

The proposal to integrate these prescalers:
* For YUV frame, apply prescalers in `pass_read_video()`, and on luma plane only. Then use existing chroma scalers to scale chroma (and possibly alpha) planes accordingly.
* For all other formats, apply prescalers to all four plane independently, after `pass_convert_yuv()`.

It's bit complicated to applied chroma scalers to match the luma plane shift, we need to adjust `chromafix` accordingly, but other than that, other parts seems to be clear. I followed @haasn's approach to handle the remaining part, passing a gl_transform `offset`, and adjust scaling parameter accordingly in `pass_scale_main()`. And as @haasn suggested, it brings enough confusion with existing `src_rect` and `dst_rect`. It might be easier to get rid of use of global parameters like `src_rect`, `dst_rect`, `image_params` in different passes, so that each pass can be reasoned independently. WDYT?

**Comment by wm4** [comment]
>It might be easier to get rid of use of global parameters like src_rect, dst_rect, image_params in different passes, so that each pass can be reasoned independently. 

The problem is that these parameters kind of have to be global, because the filter chain has to be built respecting the source and destination rectangles. But if you know of a nice way how to do that, feel free to make a suggestion.

**Comment by wm4** [comment]
By the way, would it be possible to deinterlace with nnedi? I'm not quite sure what's needed for this in vo_opengl.

**Comment by bjin** [comment]
> By the way, would it be possible to deinterlace with nnedi? I'm not quite sure what's needed for this in vo_opengl.

It should be relatively easy to use it as a deinterlacer later, it's just single pass without touch frame properties like size. But it's just for same-frame-rate deinterlacing. It's not so trivial for double-frame-rate deinterlacing (which is probably what most user want), we probably need a global option to double the framerate before frames reach the vo code (or add a customized video filter for this purpose), also a way to tag these odd and even frames.

**Comment by wm4** [comment]
>It's not so trivial for double-frame-rate deinterlacing (which is probably what most user want), we probably need a global option to double the framerate before frames reach the vo code (or add a customized video filter for this purpose), also a way to tag these odd and even frames.

Couldn't we just do the same thing as for interpolation?

**Comment by avih** [comment]
> Couldn't we just do the same thing as for interpolation?

It needs to be the fields with the original field rate (double the frame rate). IMO libavfilter should be able to separate the fields (or BOB), and from there NNEDI3 could work its scaling per field.

However, it would also need half-pixel shift up/down to match the fields spatial position, which plain field separation usually doesn't do, but IIRC the avisynth NNEDI3 filter does do.

**Comment by bjin** [comment]
> Couldn't we just do the same thing as for interpolation?

Not sure about this. As I understand, `vo` events are modelled by display refresh rate and accurate frame time. Since doubling frame rate is irrelevant to display refresh rate, it should just be easier to do it before frame reaches vo.

**Comment by wm4** [comment]
I'd want it to work with hardware decoding (which is my only motivation for it), and a software filter would break this. Anyway, deinterlacing is not the central point of this; it'd just be nice to have since I "need" it.

**Comment by haasn** [comment]
> Couldn't we just do the same thing as for interpolation?

We absolutely could, just “queue” up every frame twice, once with the even field and once with the odd field.

We have multiple queues though that we can take advantage of. I think the best place to move the explicit logic for this to would be the generic VO code that's in charge of passing frames (plus possible future frames) to the VO implementation.

To this end, I would add a new field to struct vo_frame indicating the field mode for that frame (0 = none/whole, 1 = even field only, 2 = odd field only) and the VO backend would have to respect this field by filtering appropriately.

So for example, instead of sending A B C D at timestamps 0 1 2 3, it would send A1 A2 B1 B2 C1 C2 D1 D3 at timestamps 0 0.5 1 1.5 2 2.5 3 3.5

In general, this approach would allow for “more complicated” logic than single-frame or bob deinterlacing, too, because any deinterlacer code could just look at the “next” frames and construct any complicated sequence of intermediate frames it wants to.

**Comment by wm4** [comment]
It also somehow has to work with display-sync frame scheduling, which lives in the player code, not vo.c.

**Comment by bjin** [comment]
Some followup. It seems that there are license issue to port the NNEDI3 shader. The HLSL shader I mentioned above is ported from an [OpenCL shader](http://forum.doom9.org/showthread.php?t=169766), and both of them are licensed under LGPLv3.

Since mpv is [switching to LGPL2.1](https://github.com/mpv-player/mpv/blob/master/Copyright), which is incompatible with LGPLv3. It's not possible to get non-self-authored LGPLv3 code merged and switch to LGPL2.1 later.

**Comment by haasn** [comment]
Correct me if I'm wrong, but if you reimplement an algorithm while using an existing source code as a reference to understand how it works, it is not considered a derived work.

I don't know if my implementation of super-xbr is considered sufficiently different for this to also be the case, in this case I did not make a large effort to rewrite lines in the “mpv style”, though I did still significantly change the source in the process of merging it. I'm not sure whether it would be considered a derived work or not. (If doubt arises, I could easily go through the source and replace each line by a more mpv-esque construction, but in the case of super-xbr I just didn't think it was worth the effort)

But since this is a gray area, it's probably better to be safe than sorry and just use the shader as a reference for a reimplementation that does not share any source fragments (except co-incidentally).

**Comment by bjin** [comment]
Yes, a clean reimplmentation is definitely okay. But I didn't realize this issue before and what I did for NNEDI3 shader is basically copy the HLSL one and translate it into GLSL from top line to the last. I will try to reimplement it myself, but since NNEDI3 is quite simple and straightforward (the NN weights format are also fixed), it probably won't make much difference, at least it appears so.

**Comment by bjin** [comment]
I uploaded my current progress for nnedi3 support. It's far from being ready to be merged (aside from the nnedi3 license issue). Document changes and probably some more cleans up are required. There are also issues for the performance of nnedi3, the texture approach is rather slow (compare to hardcoding all weights in shader code). See https://github.com/bjin/mpv/commits/pre-scalers

@wm4 I will open a PR once these issues are resolved, but could you do an early check first if my approach to add prescaling support is okay or not? It's mostly in the "superxbr" commit.

**Comment by wm4** [comment]
Hm, I'm not sure if I'm even suited to review this code. Maybe haasn can do this?

The source_shader leftovers should be removed in a separate commit.

The super-xbr/nnedi stuff should be in their own files each (they're large enough).

For nnedi, surely we don't need _all_ those weights.

>  // modifications by madshi (for use in madVR):

How can madVR use the shaders when they're LGPL3?

**Comment by bjin** [comment]
> Hm, I'm not sure if I'm even suited to review this code. Maybe haasn can do this?
@haasn 

> For nnedi, surely we don't need all those weights.
I will remove nns=256, this saves roughly half the size. For upscaling, probably nns=64 is enough. But for deinterlacing, nns=128 should be fine.

> How can madVR use the shaders when they're LGPL3?
My guess is that he get alternative license/permit from opencl shader author. But I don't really know.

**Comment by wm4** [comment]
Then you could ask the madVR developer for adequately licensed shader code.

**Comment by bjin** [comment]
I tried different approach to upload these weights.

1. The original approach, create a 1D texture sampler and set it as uniform (similar to how dither matrix are handled). Pretty slow (more than two times slower than the other two approach), but without any other issue.
2. Upload a `vec4[]` array as uniform, much faster, but can only handle `nns=16` and `nns=32`, there seems to be limitation on the size of uniform variables.
3. Hard code all variables into shader code, the fastest among these three approach. But the shader compiling is rather slow with mesa, essentially unusable. Also my current implementation has float number precision issue for `nns>=64` which requires an extension from OpenGL 3.3 (`ARB_shader_bit_encoding`) to address. It also have weird issue with `nns=16` alone, which I don't know the reason for now.

It seems that texture buffer and uniform buffer objects can address some of these issues for the first two approaches, but they requires OpenGL 3.1.

For now I will probably focus on the third approach to see if I can address all these issues with some workaround. But probably other approach is more promising to others. WDYT?

**Comment by wm4** [comment]
>Upload a vec4[] array as uniform, much faster, but can only handle nns=16 and nns=32, there seems to be limitation on the size of uniform variables.

I don't know, maybe it has to do with the ``GL_MAX_UNIFORM_BLOCK_SIZE`` limit?

Requiring OpenGL version greater than 3.0 is ok, as long as the functionality is optional, e.g. nnedi can simply be disabled if the GL version is not sufficient. Same for any extensions.

**Comment by bjin** [comment]
> Requiring OpenGL version greater than 3.0 is ok, as long as the functionality is optional, e.g. nnedi can simply be disabled if the GL version is not sufficient. Same for any extensions.

It's slightly different from other extensions. More specific, I need to add `#version 330` to shader if nnedi3 is enabled. But it seems that using high GLSL version have issues on some platforms, I see a workaround for OSX.
```
         // Specifically needed for OSX (normally we request 3.0 contexts only, but
         // OSX always creates 3.2 contexts when requesting a core context).     
         if (gl->version >= 320)                                                 
             gl->glsl_version = 150;  
```
Also, I'm not sure if it's okay to change `gl->gl_version` during rendering. Another option is to check vo options to see if nnedi3 is enabled or not at the GLSL version detection code above.

**Comment by wm4** [comment]
Raising the GLSL version as long as the GL version is high enough should be fine.

I'm not really sure why the OSX hack is even needed. Maybe the comment is only there because we're normally fine with GLSL 130, and don't need a higher version - just on OSX we need at least 150.

**Comment by zachsaw** [comment]
@bjin While I don't oppose to your translating of nnedi3.hlsl to GLSL, I'd like to point out that a clean room implementation isn't the same as just re-implementing the same functionality after having already translated it from HLSL to GLSL in the first place (see http://digital-law-online.info/lpdi1.0/treatise27.html). AFAICS MPV currently contains derivative works of MPDN Extensions' nnedi3.hlsl. Adding more functionalities on top of it (e.g. 8x6) is in fact the very meaning of derivative work. Our work is based on another LGPLv3.0 code by SEt, and we duly retain copyright message in our source code.

**Comment by bjin** [comment]
@zachsaw Thanks for pointing this out. To clarify, the reimplementation is done from scratch without taking any existing code as reference. It's still unclear to me wether this could be considered as derivative work or not.

I would open another PR to add all credits.

**Comment by Xuno** [comment]
Is there an archived copy of nnedi3_weights.bin available that includes nns=256 weights?

**Comment by haasn** [comment]
@bjin the current prescaling code is rather overly complex and has been getting in the way of some refactoring I'd like to do for a rather long time.

Would you be willing to reimplement prescaling (both superxbr and NNEDI3) on top of a refactored texture pipeline in a more clean way? Otherwise I would probably just cut it for now and maybe try to add it again sometime in the future.

Also, the new iteration of NNEDI3 (if any) could actually use compute shaders to avoid much of the complexity and slowness that it currently exhibits - in particular, you can use a compute shader to just process each 8x6 region as a single work group using efficient local shared memory.

**Comment by bjin** [comment]
> Would you be willing to reimplement prescaling (both superxbr and NNEDI3) on top of a refactored texture pipeline in a more clean way? Otherwise I would probably just cut it for now and maybe try to add it again sometime in the future.

@haasn Sorry, I don't have time to do this myself recently, at least in March. So if you feel it being necessary, please go ahead.
